src/app_service/service/message/dto/standard_link_preview.nim

```diff
@@ -0,0 +1,114 @@
+import json, strformat, NimQml
+import ./link_preview_thumbnail
+include ../../../common/json_utils
+
+type
+  LinkType* {.pure.} = enum
+    Link = 0
+    Image
+
+proc toLinkType*(value: int): LinkType =
+  try:
+    return LinkType(value)
+  except RangeDefect:
+    return LinkType.Link
+
+
+QtObject:
+  type StandardLinkPreview* = ref object of QObject
+    url*: string # this property is set manually and only used in`toJSON` conversion
+    hostname: string
+    title: string
+    description: string
+    linkType: LinkType
+    thumbnail: LinkPreviewThumbnail
+
+  proc setup*(self: StandardLinkPreview) =
+    self.QObject.setup
+    self.thumbnail = newLinkPreviewThumbnail()
+
+  proc delete*(self: StandardLinkPreview) =
+    self.QObject.delete
+    self.thumbnail.delete
+
+  proc newStandardLinkPreview*(hostname: string, title: string, description: string, thumbnail: LinkPreviewThumbnail, linkType: LinkType): StandardLinkPreview =
+    new(result, delete)
+    result.setup()
+    result.hostname = hostname
+    result.title = title
+    result.description = description
+    result.linkType = linkType
+    result.thumbnail.copy(thumbnail)
```

Comment:
We don't want to replace the object, as it's populated to QML by pointer.
Though I can see now that I didn't fire the change signals in `update` function. Will add that.

Reply:
Not sure I follow.

At this step the `StandardLinkPreview` ptr isn't available to qml as we're building it here. Even if it is exposed to qml probably an `thumbnailChanged` signal would do the trick. 🤔 

I was thinking it would be safe to remove the manual memory handling of `thumbnail` from `setup` and `delete` and let the GC do it's magic if we will use assignment instead of copy (1+ to less code, 1+ to less manual memory management)